service/gcs/stdio/stdio.go

```diff
@@ -99,6 +99,136 @@ func (s *ConnectionSet) Files() (_ *FileSet, err error) {
+	pr := &PipeRelay{s: s}
+	defer func() {
+		if err != nil {
+			pr.closePipes()
+		}
+	}()
+
+	if s.In != nil {
+		pr.pipes[0], pr.pipes[1], err = os.Pipe()
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create stdin pipe relay")
+		}
+	}
+	if s.Out != nil {
+		pr.pipes[2], pr.pipes[3], err = os.Pipe()
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create stdout pipe relay")
+		}
+	}
+	if s.Err != nil {
+		pr.pipes[4], pr.pipes[5], err = os.Pipe()
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create stderr pipe relay")
+		}
+	}
+	return pr, nil
+}
+
+// PipeRelay is a relay built to expose a pipe interface
+// for stdin, stdout, stderr ontop of a ConnectionSet.
+type PipeRelay struct {
+	wg sync.WaitGroup
+	s  *ConnectionSet
+	// pipes format is stdin [0 read, 1 write], stdout [2 read, 3 write], stderr [4 read, 5 write].
+	pipes [6]*os.File
+}
+
+// Files returns a FileSet with an os.File for each connection
+// in the connection set.
+func (pr *PipeRelay) Files() (*FileSet, error) {
+	fs := new(FileSet)
+	if pr.s.In != nil {
+		fs.In = pr.pipes[0]
+	}
+	if pr.s.Out != nil {
+		fs.Out = pr.pipes[3]
+	}
+	if pr.s.Err != nil {
+		fs.Err = pr.pipes[5]
+	}
+	return fs, nil
+}
+
+// Start starts the relay operation. The caller must call Wait to wait
+// for the relay to finish and release the associated resources.
+func (pr *PipeRelay) Start() {
+	if pr.s.In != nil {
+		pr.wg.Add(1)
+		go func() {
+			if _, err := io.Copy(pr.pipes[1], pr.s.In); err != nil {
+				logrus.Errorf("error copying stdin to pipe: %s", err)
+			}
+			if err := pr.pipes[1].Close(); err != nil {
+				logrus.Errorf("error closing stdin write pipe: %s", err)
+			}
+			pr.pipes[1] = nil
+			pr.wg.Done()
+		}()
+	}
+	if pr.s.Out != nil {
+		pr.wg.Add(1)
+		go func() {
+			if _, err := io.Copy(pr.s.Out, pr.pipes[2]); err != nil {
+				logrus.Errorf("error copying stdout from pipe: %s", err)
+			}
+			if err := pr.s.Out.Close(); err != nil {
+				logrus.Errorf("error closing stdout socket: %s", err)
+			}
+			pr.wg.Done()
+		}()
+	}
+	if pr.s.Err != nil {
+		pr.wg.Add(1)
+		go func() {
+			if _, err := io.Copy(pr.s.Err, pr.pipes[4]); err != nil {
+				logrus.Errorf("error copying stderr from pipe: %s", err)
+			}
+			if err := pr.s.Err.Close(); err != nil {
+				logrus.Errorf("error closing stderr socket: %s", err)
+			}
+			pr.wg.Done()
+		}()
+	}
+}
+
+// Wait waits for the relaying to finish and closes the associated
+// pipes and connections.
+func (pr *PipeRelay) Wait() {
+	// Close stdin so that the copying goroutine is safely unblocked; this is necessary
+	// because the host expects stdin to be closed before it will report process
+	// exit back to the client, and the client expects the process notification before
+	// it will close its side of stdin (which io.Copy is waiting on in the copying goroutine).
+	if pr.s.In != nil {
+		if err := pr.s.In.CloseRead(); err != nil {
```

Comment:
Maybe EndStdioAndWait or something? It could also make sense to pull the CloseRead code into either a different function, or just stick it in *process.Wait. I'm cool with any option, I think just calling it Wait implies it doesn't perform any close operations.